File: test_game_logic.js
// Test script to simulate a Doppelkopf game and verify team display and score calculation
// This script simulates the game logic without using the GUI

// Mock game state
let gameState = {
    gameId: "test-game-123",
    currentPlayer: 0,
    playerTeam: "RE",
    gameVariant: "NORMAL",
    hand: [],
    currentTrick: [],
    legalActions: [],
    scores: [120, 120], // Example scores for RE and CONTRA
    gameOver: false,
    winner: null,
    revealed_team: false,
    revealed_teams: [false, false, false, false],
    player_team_type: '',
    player_team_types: ['', '', '', ''],
    otherPlayers: [
        { id: 1, team: "CONTRA", card_count: 10, score: 0, revealed_team: false },
        { id: 2, team: "RE", card_count: 10, score: 0, revealed_team: false },
        { id: 3, team: "CONTRA", card_count: 10, score: 0, revealed_team: false }
    ]
};

// Mock cards
const mockCards = {
    queenOfClubs: { id: "QC1", rank: "QUEEN", suit: "CLUBS", is_second: false },
    aceOfSpades: { id: "AS1", rank: "ACE", suit: "SPADES", is_second: false },
    tenOfHearts: { id: "10H1", rank: "TEN", suit: "HEARTS", is_second: false },
    jackOfDiamonds: { id: "JD1", rank: "JACK", suit: "DIAMONDS", is_second: false }
};

// Test functions
function testTeamReveal() {
    console.log("=== Testing Team Reveal Logic ===");
    
    // Initial state - no teams revealed
    console.log("Initial state:");
    console.log("Player revealed team:", gameState.revealed_team);
    console.log("Player team type:", gameState.player_team_type);
    console.log("Other players revealed teams:", gameState.revealed_teams);
    console.log("Other players team types:", gameState.player_team_types);
    
    // Simulate player playing Queen of Clubs
    console.log("\nPlayer plays Queen of Clubs:");
    
    // This simulates the logic in playCard function
    gameState.revealed_team = true;
    gameState.player_team_type = 're';
    
    // Count how many "re" players we've identified
    let reCount = 1; // We just identified ourselves
    
    if (gameState.player_team_types) {
        for (let i = 0; i < gameState.player_team_types.length; i++) {
            if (gameState.player_team_types[i] === 're') {
                reCount++;
            }
        }
    }
    
    console.log("After player plays Queen of Clubs:");
    console.log("Player revealed team:", gameState.revealed_team);
    console.log("Player team type:", gameState.player_team_type);
    console.log("Other players revealed teams:", gameState.revealed_teams);
    console.log("Other players team types:", gameState.player_team_types);
    
    // Simulate AI player 2 playing Queen of Clubs
    console.log("\nAI Player 2 plays Queen of Clubs:");
    
    // This simulates the logic in renderCurrentTrick function
    const playerIdx = 2;
    
    // Initialize the revealed_teams array if it doesn't exist
    if (!gameState.revealed_teams) {
        gameState.revealed_teams = [false, false, false, false];
        gameState.player_team_types = ['', '', '', ''];
    }
    
    gameState.revealed_teams[playerIdx] = true;
    gameState.player_team_types[playerIdx] = 're';
    
    // Count how many "re" players we've identified
    reCount = 0;
    if (gameState.player_team_type === 're') {
        reCount++;
    }
    
    if (gameState.player_team_types) {
        for (let i = 0; i < gameState.player_team_types.length; i++) {
            if (gameState.player_team_types[i] === 're') {
                reCount++;
            }
        }
    }
    
    // If we've identified both "re" players, mark the others as "contra"
    if (reCount === 2) {
        // For the human player
        if (gameState.player_team_type !== 're') {
            gameState.revealed_team = true;
            gameState.player_team_type = 'contra';
        }
        
        // For AI players
        if (gameState.player_team_types) {
            for (let i = 0; i < gameState.player_team_types.length; i++) {
                if (gameState.player_team_types[i] !== 're') {
                    gameState.revealed_teams[i] = true;
                    gameState.player_team_types[i] = 'contra';
                }
            }
        }
    }
    
    console.log("After AI player 2 plays Queen of Clubs:");
    console.log("Player revealed team:", gameState.revealed_team);
    console.log("Player team type:", gameState.player_team_type);
    console.log("Other players revealed teams:", gameState.revealed_teams);
    console.log("Other players team types:", gameState.player_team_types);
    
    // Verify that all teams are now revealed
    const allTeamsRevealed = gameState.revealed_team && 
                            gameState.revealed_teams.every(revealed => revealed);
    
    console.log("\nAll teams revealed:", allTeamsRevealed);
    
    // Verify that we have 2 re players and 2 contra players
    // Note: We need to be careful about double-counting the player
    let rePlayerCount = 0;
    let contraPlayerCount = 0;
    
    // Count the human player
    if (gameState.player_team_type === 're') {
        rePlayerCount++;
    } else if (gameState.player_team_type === 'contra') {
        contraPlayerCount++;
    }
    
    // Count the AI players (only the first 3 since we have 4 players total)
    for (let i = 1; i <= 3; i++) {
        if (gameState.player_team_types[i] === 're') {
            rePlayerCount++;
        } else if (gameState.player_team_types[i] === 'contra') {
            contraPlayerCount++;
        }
    }
    
    console.log("Re player count:", rePlayerCount);
    console.log("Contra player count:", contraPlayerCount);
    
    return rePlayerCount === 2 && contraPlayerCount === 2;
}

function testScoreCalculation() {
    console.log("\n=== Testing Score Calculation Logic ===");
    
    // Set up game over state
    gameState.gameOver = true;
    gameState.winner = "RE"; // RE team wins
    
    // Calculate the points for each player based on the winner
    const winningTeam = gameState.winner;
    const pointsPerPlayer = 2; // +2 for winners, -2 for losers
    
    // Calculate scores for each player
    const playerScores = [];
    
    // For player 0 (human player)
    const playerTeam = gameState.playerTeam;
    const playerPoints = playerTeam === winningTeam ? pointsPerPlayer : -pointsPerPlayer;
    playerScores.push({ id: 0, team: playerTeam, points: playerPoints });
    
    // For AI players
    for (let i = 0; i < gameState.otherPlayers.length; i++) {
        const player = gameState.otherPlayers[i];
        const points = player.team === winningTeam ? pointsPerPlayer : -pointsPerPlayer;
        playerScores.push({ id: player.id, team: player.team, points: points });
    }
    
    // Log the results
    console.log("Game over, winning team:", winningTeam);
    console.log("Player scores:");
    playerScores.forEach(player => {
        console.log(`Player ${player.id} (${player.team}): ${player.points} points`);
    });
    
    // Verify that we have 2 players with +2 points and 2 players with -2 points
    const winnersCount = playerScores.filter(player => player.points === pointsPerPlayer).length;
    const losersCount = playerScores.filter(player => player.points === -pointsPerPlayer).length;
    
    console.log("\nPlayers with +2 points:", winnersCount);
    console.log("Players with -2 points:", losersCount);
    
    return winnersCount === 2 && losersCount === 2;
}

// Run the tests
console.log("Starting Doppelkopf game logic tests...\n");

const teamRevealSuccess = testTeamReveal();
const scoreCalculationSuccess = testScoreCalculation();

console.log("\n=== Test Results ===");
console.log("Team reveal logic:", teamRevealSuccess ? "PASSED" : "FAILED");
console.log("Score calculation logic:", scoreCalculationSuccess ? "PASSED" : "FAILED");
console.log("All tests:", (teamRevealSuccess && scoreCalculationSuccess) ? "PASSED" : "FAILED");

File: test_game_logic.py
#!/usr/bin/env python3
"""
Test script to verify the team display and score calculation logic in the Doppelkopf game.
"""

from game.doppelkopf import DoppelkopfGame, Card, Suit, Rank, GameVariant, PlayerTeam

def test_team_reveal():
    """Test the team reveal logic when a Queen of Clubs is played."""
    print("=== Testing Team Reveal Logic ===")
    
    # Create a new game
    game = DoppelkopfGame()
    
    # Reset the game with a fixed seed for reproducibility
    import random
    random.seed(42)
    game.reset()
    
    # Print initial teams (should be determined but not revealed to players)
    print("Initial teams:")
    for i, team in enumerate(game.teams):
        print(f"Player {i}: {team.name}")
    
    # Find a player with a Queen of Clubs
    queen_of_clubs = Card(Suit.CLUBS, Rank.QUEEN, False)
    queen_of_clubs2 = Card(Suit.CLUBS, Rank.QUEEN, True)
    
    re_players = []
    for i, hand in enumerate(game.hands):
        if queen_of_clubs in hand or queen_of_clubs2 in hand:
            re_players.append(i)
            print(f"Player {i} has a Queen of Clubs and is on team RE")
    
    # There should be exactly 2 players on team RE
    assert len(re_players) == 2, f"Expected 2 RE players, found {len(re_players)}"
    
    # Verify that the teams are correctly assigned
    for i, team in enumerate(game.teams):
        if i in re_players:
            assert team == PlayerTeam.RE, f"Player {i} should be on team RE"
        else:
            assert team == PlayerTeam.KONTRA, f"Player {i} should be on team KONTRA"
    
    print("Team assignment test passed!")
    
    # In a real game, when a Queen of Clubs is played, it would reveal the player's team
    # This is handled in the web interface (game.js) but we can simulate it here
    
    # Simulate a game where a Queen of Clubs is played
    # First, make sure the player with Queen of Clubs is the current player
    re_player = re_players[0]
    game.current_player = re_player
    
    # Find the Queen of Clubs in their hand
    queen_in_hand = None
    for card in game.hands[re_player]:
        if card.rank == Rank.QUEEN and card.suit == Suit.CLUBS:
            queen_in_hand = card
            break
    
    assert queen_in_hand is not None, f"Player {re_player} should have a Queen of Clubs"
    
    # Play the Queen of Clubs
    print(f"\nPlayer {re_player} plays Queen of Clubs")
    game.play_card(re_player, queen_in_hand)
    
    # The current trick should now contain the Queen of Clubs
    assert game.current_trick[0] == queen_in_hand, "Queen of Clubs should be in the current trick"
    
    print("Queen of Clubs played successfully!")
    
    # In the web interface, this would trigger the team display to show "re" for this player
    # We can verify that the player is indeed on team RE
    assert game.teams[re_player] == PlayerTeam.RE, f"Player {re_player} should be on team RE"
    
    print("Team reveal test passed!")
    return True

def test_score_calculation():
    """Test the score calculation logic at the end of a game."""
    print("\n=== Testing Score Calculation Logic ===")
    
    # Create a new game
    game = DoppelkopfGame()
    
    # Reset the game with a fixed seed for reproducibility
    import random
    random.seed(42)
    game.reset()
    
    # Print initial teams
    print("Teams:")
    for i, team in enumerate(game.teams):
        print(f"Player {i}: {team.name}")
    
    # Simulate playing all cards to end the game
    # We'll just set the scores directly and call _end_game
    
    # Set scores so that RE team wins (needs at least 121 points)
    game.scores = [150, 90]  # [RE score, KONTRA score]
    
    # End the game
    game._end_game()
    
    # Verify that the game is over and RE team won
    assert game.game_over, "Game should be over"
    assert game.winner == PlayerTeam.RE, "RE team should win"
    
    print(f"Game over, winner: {game.winner.name}")
    
    # In the web interface, this would trigger the score display
    # The winning team gets +2 points per player, losing team gets -2 points per player
    
    # Count players in each team
    re_players = sum(1 for team in game.teams if team == PlayerTeam.RE)
    kontra_players = sum(1 for team in game.teams if team == PlayerTeam.KONTRA)
    
    assert re_players == 2, f"Expected 2 RE players, found {re_players}"
    assert kontra_players == 2, f"Expected 2 KONTRA players, found {kontra_players}"
    
    # Calculate expected points
    re_points = kontra_players  # Each RE player gets +kontra_players
    kontra_points = -re_players  # Each KONTRA player gets -re_players
    
    print(f"RE players should get {re_points} points each")
    print(f"KONTRA players should get {kontra_points} points each")
    
    # Verify that the points would be calculated correctly
    for i, team in enumerate(game.teams):
        expected_points = re_points if team == PlayerTeam.RE else kontra_points
        print(f"Player {i} ({team.name}): Expected {expected_points} points")
    
    print("Score calculation test passed!")
    return True

def main():
    """Run the tests."""
    print("Starting Doppelkopf game logic tests...\n")
    
    team_reveal_success = test_team_reveal()
    score_calculation_success = test_score_calculation()
    
    print("\n=== Test Results ===")
    print(f"Team reveal logic: {'PASSED' if team_reveal_success else 'FAILED'}")
    print(f"Score calculation logic: {'PASSED' if score_calculation_success else 'FAILED'}")
    print(f"All tests: {'PASSED' if team_reveal_success and score_calculation_success else 'FAILED'}")

if __name__ == "__main__":
    main()
